### PR TITLE
[FIXES JENKINS-49500] Prevent triggers from being restarted if same

### PR DESCRIPTION
**Pull request by scoheb**
* JENKINS issue(s):
    * https://issues.jenkins-ci.org/browse/JENKINS-49500
* Description:
    * Prevent triggers from being restarted if they have not changed between builds.

### CODE REVIEW

**Comment by scoheb** [comment]
@abayer for your review.

**Comment by bitwiseman** [comment]
@scoheb  I'll probably be the one to review this.  Could you give a little more details about this?

**Comment by scoheb** [comment]
@bitwiseman I will describe my use case for this....

We use the jms-messaging plugin for our triggers for our pipelines. The plugin creates a thread that subscribes to a topic on a message broker. If the declarative pipeline contains a `triggers` section, then every single time a build starts up, the trigger is blindly stopped/started (even though no changes were made to the trigger configuration). Since we use a thread to monitor for new messages, then we are constantly and needlessly destroying/creating threads and there is a risk that we even miss a message since we might not be subscribed since we are stopping/starting.  

This PR attempts to detect that the trigger has not been changed and thus it prevents the stop/start that can be disruptive.

hope that helps clarify the issue!

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 1]
```diff
@@ -630,17 +631,24 @@ class Utils {
                 }
             }
 
-            // Remove the triggers/parameters properties regardless.
-            j.removeProperty(PipelineTriggersJobProperty.class)
+            def currentTriggerProp = j.getProperty(PipelineTriggersJobProperty.class)
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: All the triggers are in one property, but there can be multiple, so plural. 
```suggestion
            def currentTriggersProp = j.getProperty(PipelineTriggersJobProperty.class)
```

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 2]
```diff
@@ -630,17 +631,24 @@ class Utils {
...
+            def newTriggerProp = new PipelineTriggersJobProperty(triggersToApply)
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]:
```suggestion
            def newTriggersProp = new PipelineTriggersJobProperty(triggersToApply)
```

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 3]
```diff
@@ -630,17 +631,24 @@ class Utils {
...
+            // Only Remove/Add if the triggers are different.
+            if (!areTriggersInPipelineTriggerJobPropertiesEqual(currentTriggerProp, newTriggerProp)) {
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]:
```suggestion
            if (!areTriggersInPipelineTriggerJobPropertiesEqual(currentTriggersProp, newTriggersProp)) {
```

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 4]
```diff
@@ -630,17 +631,24 @@ class Utils {
...
+                j.removeProperty(PipelineTriggersJobProperty.class)
+                if (!triggersToApply.isEmpty()) {
+                    j.addProperty(newTriggerProp)
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]:
```suggestion
                    j.addProperty(newTriggersProp)
```

Also, this replaces all the triggers even if only one has changed, right?
It would be better if made it so only the _changed_ triggers are replaced.  But that might be work for another PR.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 5]
```diff
@@ -663,6 +671,14 @@ class Utils {
         }
     }
 
+    // Utility method to determine is triggers in PipelineTriggersJobProperties are equal
+    static boolean areTriggersInPipelineTriggerJobPropertiesEqual(PipelineTriggersJobProperty left, PipelineTriggersJobProperty right) {
+        if (left == right) return true;
+        if (left == null || right == null) return false;
+        boolean triggersEqual = CollectionUtils.isEqualCollection(left.triggers, right.triggers);
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: Is this comparison unordered?  Will it return true even if only the order of the triggers has changed?

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/TriggersTest.java [hunk 6]
```diff
@@ -162,4 +176,113 @@ public void actualTriggerCorrectScope() throws Exception {
         expect("actualTriggerCorrectScope")
                 .go();
     }
+
+    @LocalData
+    @Test
+    public void doNotRestartEqualTriggers() throws Exception {
```
Comment (bitwiseman) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/TriggersTest.java]: This is good. 
We also need a test to show this with two triggers and that if the order of the triggers changes (but they are otherwise the same) that this still works and does not recreate the triggers.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy (hunk 5 above)
Comment (gulyaev13) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @bitwiseman , @scoheb , sorry for interrupting your review, but I think this will not work.
`CollectionUtils.isEqualCollection()` use hashCode for comparing objects, but Trigger and child classes don't override this method, so it will not work.

But the problem is more complex, than the problem with Triggers. 
The whole method `org.jenkinsci.plugins.pipeline.modeldefinitionUtils.updateJobProperties` should be refactored, because it isn't thread-safe (I can be one of the reasons of [JENKINS-44809](https://issues.jenkins-ci.org/browse/JENKINS-44809)) and affect performance in case of parallel pipeline execution (save configuration each time, even if it has not been changed).

I'm working on a solution, which will resolve performance problems.
I want to add some hints, based on specifics of tipis of JobProperties. 
But exactly for triggers, I can't find a solution better than comparing Strings after XSTREAM serialization. Maybe you can advise me something before PR?

Comment (bitwiseman) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @gulyaev13 
Interesting. Given what you're saying with hash code I wonder how the test succeeded? 

As to multi-threading - that would mean the the current running implementation is also not thread-safe.  Is what you have seen and are trying to fix?

Comment (gulyaev13) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @bitwiseman , there is no magic. TestTrigger defined in the test has overwritten methods hashCode() and equals().

Yes, the current implementation isn't seen thread-safe, because of method updateJobProperties change global object WorkflowJob without any sync. 
I've seen the effects of concurrent changing a few times ago. The job's parameters were deleted from one thread and were saved to disk by another thread.
The main problem which I'm trying to solve is reducing a lot of IO calls caused by saving same configuration each job run (Save action is thread-save and become a bottleneck of concurrent pipeline execution).

Comment (scoheb) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @gulyaev13 @bitwiseman I think the PR still improves the situation for triggers that do implement `equals()` and `hashcode()`

Comment (gulyaev13) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @scoheb, I think I'm ready to send my PR which will solve your problem, too!
Am I add your test to my PR?

Comment (scoheb) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: @gulyaev13 of course you can use my test! thanks! looking forward to your PR!

**Comment by bitwiseman** [comment]
@scoheb @gulyaev13 
Please coordinate your efforts here if you can.  @scoheb I assume you'll want to close this change and use #365 ?  Would you be willing to write some more test for #365, perhaps?

**Comment by scoheb** [comment]
@bitwiseman yes, I will close this in favour of #365
